Test works normally but fails under pytest with assertion rewriting

Given a test file `p.py` containing:
```python
def test_add():
    add = lambda *t: sum(t)
    l = range(8)
    e = iter(l)
    assert sum(l[:4]) == add(*[next(e) for j in range(4)])
```
the test doesn't work under pytest with assertion rewriting. Note that the expression on the rhs of `==` will return a different result when evaluated a second time. The first time it is evaluated it will give `6` and the second time `22`.

The test passes (as expected) when imported normally:
```python
>>> import p
>>> p.test_add()
```

It also works under pytest without assertion rewriting
```console
$ pytest p.py --assert=plain
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.6.2, pytest-3.10.0, py-1.7.0, pluggy-0.8.0
rootdir: /Users/enojb/current/sympy/sympy, inifile:
plugins: faulthandler-1.5.0
collected 1 item                                                                                                                                             

p.py .                                                                                                                                                 [100%]

================================================================== 1 passed in 0.01 seconds ==================================================================
```

However it fails under pytest with assertion rewriting
```console
$ pytest p.py
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.6.2, pytest-3.10.0, py-1.7.0, pluggy-0.8.0
rootdir: /Users/enojb/current/sympy/sympy, inifile:
plugins: faulthandler-1.5.0
collected 1 item                                                                                                                                             

p.py F                                                                                                                                                 [100%]

========================================================================== FAILURES ==========================================================================
__________________________________________________________________________ test_add __________________________________________________________________________

    def test_add():
        add = lambda *t: sum(t)
        l = range(8)
        e = iter(l)
>       assert sum(l[:4]) == add(*[next(e) for j in range(4)])
E       assert 6 == 22
E        +  where 6 = sum(range(0, 4))
E        +  and   22 = <function test_add.<locals>.<lambda> at 0x10454e048>(*[0, 1, 2, 3])

p.py:6: AssertionError
================================================================== 1 failed in 0.08 seconds ==================================================================
```

--- a/changelog/4412.bugfix.rst
+++ b/changelog/4412.bugfix.rst
@@ -1,0 +1,1 @@
+Fix assertion rewriting involving ``Starred`` + side-effects.

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -946,7 +946,8 @@ class AssertionRewriter(ast.NodeVisitor):
     def visit_Starred(self, starred):
         # From Python 3.5, a Starred node can appear in a function call
         res, expl = self.visit(starred.value)
-        return starred, "*" + expl
+        new_starred = ast.Starred(res, starred.ctx)
+        return new_starred, "*" + expl
 
     def visit_Call_legacy(self, call):
         """

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -413,6 +413,19 @@ class TestAssertionRewrite(object):
         )
         testdir.runpytest().assert_outcomes(passed=1)
 
+    @pytest.mark.skipif("sys.version_info < (3,5)")
+    def test_starred_with_side_effect(self, testdir):
+        """See #4412"""
+        testdir.makepyfile(
+            """\
+            def test():
+                f = lambda x: x
+                x = iter([1, 2, 3])
+                assert 2 * next(x) == f(*[next(x)])
+            """
+        )
+        testdir.runpytest().assert_outcomes(passed=1)
+
     def test_call(self):
         def g(a=42, *args, **kwargs):
             return False

